app/services/catalog/copy_portfolio.rb

```diff
@@ -0,0 +1,33 @@
+module Catalog
+  class CopyPortfolio
+    attr_reader :new_portfolio
+
+    def initialize(params)
+      @name = params[:portfolio_name]
+      @portfolio = Portfolio.find(params[:portfolio_id])
+    end
+
+    def process
+      @new_portfolio = make_copy
+
+      self
+    end
+
+    private
+
+    def make_copy
+      @portfolio.dup.tap do |new_portfolio|
+        new_portfolio.name = @name || Catalog::NameAdjust.create_copy_name(@portfolio.name, Portfolio.all.pluck(:name))
+        new_portfolio.save
```

Comment:
@lindgrenj6 If the name is already in use we have to let the user know that the name is already in use so they can provide a different one. This is only true when the user has provided a name otherwise the NameAdjust will create the nice name. If the user wants us to create the name they would leave the field empty.
I am not sure how the default save throws the name duplicate error and if its evident to the user which parameter is incorrect. It might just work but test it.

Reply:
@lindgrenj6 Where is Catalog::NameAdjust defined?

Reply:
I can change it to a `.save!` that way it will throw an exception if the name is a duplicate (whether it's user defined or not) since we have a validation on it. 

`Catalog::NameAdjust` is defined in `lib/catalog/name_adjust.rb`